### ISSUE TITLE
undefined method delete_role/delete_member when kicking a bot

### ISSUE THREAD

**Issue by localhostdotdev** [discordrb/discordrb#619]
# Summary

How to reproduce:

- Have a bot
- `bot.run(:async)`
- Add to a server via permissions url: `https://discordapp.com/oauth2/authorize/?client_id=...`
- Kick bot

```
[ERROR : websocket @ 2019-03-31 14:51:40.565] Exception: #<NoMethodError: undefined method `delete_role' for nil:NilClass>
[ERROR : websocket @ 2019-03-31 14:51:40.565] /.rvm/gems/ruby-head/gems/discordrb-3.3.0/lib/discordrb/bot.rb:905:in `delete_guild_role'
[ERROR : websocket @ 2019-03-31 14:51:40.566] /.rvm/gems/ruby-head/gems/discordrb-3.3.0/lib/discordrb/bot.rb:1200:in `handle_dispatch'
[ERROR : websocket @ 2019-03-31 14:51:40.566] /.rvm/gems/ruby-head/gems/discordrb-3.3.0/lib/discordrb/bot.rb:654:in `dispatch'
[ERROR : websocket @ 2019-03-31 14:51:40.566] /.rvm/gems/ruby-head/gems/discordrb-3.3.0/lib/discordrb/gateway.rb:714:in `handle_dispatch'
[ERROR : websocket @ 2019-03-31 14:51:40.567] /.rvm/gems/ruby-head/gems/discordrb-3.3.0/lib/discordrb/gateway.rb:679:in `handle_message'
[ERROR : websocket @ 2019-03-31 14:51:40.567] /.rvm/gems/ruby-head/gems/discordrb-3.3.0/lib/discordrb/gateway.rb:617:in `websocket_loop'
[ERROR : websocket @ 2019-03-31 14:51:40.567] /.rvm/gems/ruby-head/gems/discordrb-3.3.0/lib/discordrb/gateway.rb:569:in `connect'
[ERROR : websocket @ 2019-03-31 14:51:40.567] /.rvm/gems/ruby-head/gems/discordrb-3.3.0/lib/discordrb/gateway.rb:463:in `block in connect_loop'
[ERROR : websocket @ 2019-03-31 14:51:40.567] /.rvm/gems/ruby-head/gems/discordrb-3.3.0/lib/discordrb/gateway.rb:462:in `loop'
[ERROR : websocket @ 2019-03-31 14:51:40.568] /.rvm/gems/ruby-head/gems/discordrb-3.3.0/lib/discordrb/gateway.rb:462:in `connect_loop'
[ERROR : websocket @ 2019-03-31 14:51:40.568] /.rvm/gems/ruby-head/gems/discordrb-3.3.0/lib/discordrb/gateway.rb:168:in `block in run_async'
[ERROR : websocket @ 2019-03-31 14:51:40.739] Gateway message error!
[ERROR : websocket @ 2019-03-31 14:51:40.741] Exception: #<NoMethodError: undefined method `delete_member' for nil:NilClass>
[ERROR : websocket @ 2019-03-31 14:51:40.741] /.rvm/gems/ruby-head/gems/discordrb-3.3.0/lib/discordrb/bot.rb:854:in `delete_guild_member'
[ERROR : websocket @ 2019-03-31 14:51:40.742] /.rvm/gems/ruby-head/gems/discordrb-3.3.0/lib/discordrb/bot.rb:1175:in `handle_dispatch'
[ERROR : websocket @ 2019-03-31 14:51:40.742] /.rvm/gems/ruby-head/gems/discordrb-3.3.0/lib/discordrb/bot.rb:654:in `dispatch'
[ERROR : websocket @ 2019-03-31 14:51:40.742] /.rvm/gems/ruby-head/gems/discordrb-3.3.0/lib/discordrb/gateway.rb:714:in `handle_dispatch'
[ERROR : websocket @ 2019-03-31 14:51:40.742] /.rvm/gems/ruby-head/gems/discordrb-3.3.0/lib/discordrb/gateway.rb:679:in `handle_message'
[ERROR : websocket @ 2019-03-31 14:51:40.742] /.rvm/gems/ruby-head/gems/discordrb-3.3.0/lib/discordrb/gateway.rb:617:in `websocket_loop'
[ERROR : websocket @ 2019-03-31 14:51:40.743] /.rvm/gems/ruby-head/gems/discordrb-3.3.0/lib/discordrb/gateway.rb:569:in `connect'
[ERROR : websocket @ 2019-03-31 14:51:40.743] /.rvm/gems/ruby-head/gems/discordrb-3.3.0/lib/discordrb/gateway.rb:463:in `block in connect_loop'
[ERROR : websocket @ 2019-03-31 14:51:40.743] /.rvm/gems/ruby-head/gems/discordrb-3.3.0/lib/discordrb/gateway.rb:462:in `loop'
[ERROR : websocket @ 2019-03-31 14:51:40.743] /.rvm/gems/ruby-head/gems/discordrb-3.3.0/lib/discordrb/gateway.rb:462:in `connect_loop'
[ERROR : websocket @ 2019-03-31 14:51:40.743] /.rvm/gems/ruby-head/gems/discordrb-3.3.0/lib/discordrb/gateway.rb:168:in `block in run_async'
```

---

## Environment

**Ruby version:** ruby 2.7.0dev (2019-03-18 trunk 67296) [x86_64-darwin17]
**Discordrb version:** discordrb (3.3.0)

**Comment by localhostdotdev** [comment]
@z64 I can reproduce it with:

```ruby
require 'discordrb'
bot = Discordrb::Bot.new(token: TOKEN)
bot.run(:async)
```

**Comment by z64** [comment]
I would be impressed, since that bot just exits immediately:) I would check you understand [what passing an argument to run is for](https://www.rubydoc.info/gems/discordrb/Discordrb/Bot#run-instance_method)

Anyways; this is a known caching "bug". What happens is, when you kick the bot (with an associated managed role), you are sent three events:

1. `GUILD_DELETE`
2. `ROLE_DELETE`
3. `GUILD_MEMBER_REMOVE`

They are sent from Discord regardless that the bot is being removed, and may be received in any order.
So what happens is, if they are received in the order I listed:

1. The server is removed from our cache - this means we let go of all associated roles, members, etc.
2. The role delete event is handled and unsafely assumes a server was found in the cache and tries to call  a method to remove the role from memory
3. The member remove event is handled and unsafely assumes a server was found in the cache and tries to call  a method to remove the role from memory

This error is harmless and can be safely ignored. We can tidy the logic at some point.

**Comment by strelokhalfer** [comment]
Same with **reaction_add** and **reaction_remove**, if i try use **event.user**

**Comment by z64** [comment]
Fixed in #663
